crates/runtime/src/traphandlers/unix.rs

```diff
@@ -138,6 +138,23 @@ unsafe fn get_pc(cx: *mut libc::c_void) -> *const u8 {
         } else if #[cfg(all(any(target_os = "linux", target_os = "android"), target_arch = "aarch64"))] {
             let cx = &*(cx as *const libc::ucontext_t);
             cx.uc_mcontext.pc as *const u8
+        } else if #[cfg(all(target_os = "linux", target_arch = "s390x"))] {
+            // On s390x, SIGILL and SIGFPE are delivered with the PSW address
+            // pointing *after* the faulting instruction, while SIGSEGV and
+            // SIGBUS are delivered with the PSW address pointing *to* the
+            // faulting instruction.  To handle this, the code generator registers
+            // any trap that results in one of "late" signals on the last byte
+            // of the instruction, and any trap that results in one of the "early"
+            // signals on the first byte of the instruction (as usual).  This
+            // means we simply need to decrement the reported PSW address by
+            // one in the case of a "late" signal here to ensure we always
+            // correctly find the associated trap handler.
```

Comment:
Is there an online link to gist here which goes into some more detail? I'm mostly just curious for my own edification!

Reply:
That's a good question.  Right now, I don't think this is documented anywhere but by reading the Linux kernel sources.  This should probably be added to the ABI document.   I've added an issue there as a reminder: https://github.com/IBM/s390x-abi/issues/2

Reply:
Ok, sounds reasonable to me!